lib/utils/imagePreload.ts

```diff
@@ -0,0 +1,24 @@
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import PQueue from 'p-queue'
+
+const queue = new PQueue({ concurrency: 5 })
+
+/**
+ * Preload an image URL
+ * @param url URL of the image
+ */
+export function preloadImage(url: string): Promise<boolean> {
+	const { resolve, promise } = Promise.withResolvers<boolean>()
+	queue.add(() => {
+		const image = new Image()
+		image.onerror = () => resolve(false)
+		image.onload = () => resolve(true)
+		image.src = url
+	})
```

Comment:
queued callback itself is not waiting for preview to be fetched, so they're executed all at once, killing the efforts to rate-limit them =)

Works for me with adding one more layer of Promises (compare Network waterfall in both cases):
```suggestion
	queue.add(async () => {
		const image = new Image()
		const imagePromise = new Promise<boolean>((imgResolve) => {
			image.onload = () => imgResolve(true)
			image.onerror = () => imgResolve(false)
		})
		image.src = url
		const result = await imagePromise
		resolve(result)
	})
```

Reply:
Shall we also clear the queue on FilePicker unmount?

Reply:
```suggestion
	queue.add(() => {
		const image = new Image()
		image.onerror = () => resolve(false)
		image.onload = () => resolve(true)
		image.src = url
		return promise
	})
```